base/asyncmap.jl

```diff
@@ -212,34 +212,26 @@ end
 
 function start_worker_task!(worker_tasks, exec_func, chnl, batch_size=nothing)
     t = @async begin
-        retval = nothing
-
-        try
-            if isa(batch_size, Number)
-                while isopen(chnl)
-                    # The mapping function expects an array of input args, as it processes
-                    # elements in a batch.
-                    batch_collection=Any[]
-                    n = 0
-                    for exec_data in chnl
-                        push!(batch_collection, exec_data)
-                        n += 1
-                        (n == batch_size) && break
-                    end
-                    if n > 0
-                        exec_func(batch_collection)
-                    end
-                end
-            else
+        if isa(batch_size, Number)
+            while isopen(chnl)
+                # The mapping function expects an array of input args, as it processes
+                # elements in a batch.
+                batch_collection=Any[]
+                n = 0
                 for exec_data in chnl
-                    exec_func(exec_data...)
+                    push!(batch_collection, exec_data)
+                    n += 1
+                    (n == batch_size) && break
+                end
+                if n > 0
+                    exec_func(batch_collection)
                 end
             end
-        catch e
-            close(chnl)
```

Comment:
At a quick skim, maybe you get deadlock because this `close` has been deleted? You might need to put this back in inside a `try  ... finally` to preserve fail-fast cancellation.

Reply:
Yes, that appears to be it. Maybe `poptaskref` should fail instead of deadlocking even without `close(chnl)`? But I suppose that is outside the scope of this PR.

I have added a regression test for this deadlock.